digicampipe/io/zfits.py

```diff
@@ -35,115 +34,173 @@ def zfits_event_source(
+                            lambda x: x in data.r0.tels_with_data, sublist))
+                    for sublist in allowed_tels
+                ]
+
+            _sort_ids = None
+
+            for tel_id in data.r0.tels_with_data:
+
+                pixel_ids = event.hiGain.waveforms.pixelsIndices
+                n_pixels = len(pixel_ids)
+                if _sort_ids is None:
+                    _sort_ids = np.argsort(pixel_ids)
+                samples = event.hiGain.waveforms.samples.reshape(n_pixels, -1)
+
+                try:
+                    unsorted_baseline = event.hiGain.waveforms.baselines
+                except AttributeError:
+                    warnings.warn((
+                        "Could not read `hiGain.waveforms.baselines`"
+                        " for event:{0}"
+                        "of file:{url}".format(event_counter, url)
+                        ))
+                    return np.ones(n_pixels) * np.nan
+
+                data.inst.num_channels[tel_id] = event.num_gains
+                data.inst.geom[tel_id] = camera.geometry
+                data.inst.cluster_matrix_7[tel_id] = camera.cluster_7_matrix
+                data.inst.cluster_matrix_19[tel_id] = camera.cluster_19_matrix
+                data.inst.patch_matrix[tel_id] = camera.patch_matrix
+                data.inst.num_pixels[tel_id] = samples.shape[0]
+                data.inst.num_samples[tel_id] = samples.shape[1]
+
+                r0 = data.r0.tel[tel_id]
+                r0.camera_event_number = event.eventNumber
+                r0.pixel_flags = event.pixels_flags[_sort_ids]
+                r0.local_camera_clock = (
+                    event.local_time_sec * 1E9 + event.local_time_nanosec
+                )
+                r0.gps_time = (
+                    event.trig.timeSec * 1E9 + event.trig.timeNanoSec
+                )
+                r0.camera_event_type = event.event_type
+                r0.array_event_type = event.eventType
+                r0.adc_samples = samples[_sort_ids]
+
+                r0.trigger_input_traces = _prepare_trigger_input(
+                    event.trigger_input_traces)
+                r0.trigger_output_patch7 = _prepare_trigger_output(
+                    event.trigger_output_patch7)
+                r0.trigger_output_patch19 = _prepare_trigger_output(
+                    event.trigger_output_patch19)
+                r0.digicam_baseline = unsorted_baseline[_sort_ids]
+
+            yield data
 
 
 def count_number_events(file_list):
-    """return sum of events in all files
-    It is useful for fixed size array creation.
-
-    :param file_list: list of zfits files
-    :return: n_events: int
-    """
-    n_events = 0
-
-    for filename in file_list:
-
-        zfits = ZFile(filename)
-        n_events += zfits.numrows
-
-    return n_events
+    return sum(
+        len(SimpleFile(filename).Events)
+        for filename in file_list
+    )
+
+
+PATCH_ID_INPUT = [
```

Comment:
if we go for remapping here rather that in the protozfits package then this should be in `utils.Digicam`